### PR TITLE
Don't crash on windows

### PR DESCRIPTION
**Pull request by kohenkatz**
As discussed in #44, #40, and #21, it is currently impossible to build code on Windows that references tableflip, even if that code never actually runs tableflip.

This PR addresses that issue in two ways:

1. Allow the tableflip code to compile without errors on Windows.
  Code in `fds.go` that uses `syscall` is separated into conditionally-compiled files `dup_fd.go` and `dup_fd_windows.go`. This follows the existing pattern in `dup_file.go` and `dup_file_legacy.go`. Similarly, code in `env.go` which uses `syscall` is separated into conditionally-compiled files `env_syscalls.go` and `env_windows.go`.
  It is important to note that **the code will not work** even though it compiled successfully. That is why the next part is needed.
2. Allow Windows users to test/run code that was built using tableflip with minimal changes.
  `Upgrader` is modified to return a `tableflip.ErrNotSupported` error when running on Windows.
  Additionally, a new package `github.com/cloudflare/tableflip/testing` is provided that contains a simple stub implementation of the `Upgrader` and `Fds` classes. As shown in the example provided therein, the user can define an interface with the required methods, then check the result of `tableflip.New` to see if `errors.Is(err, tableflip.ErrNotSupported`. If that is the case, the user can create an instance of the stub and use that instead.

### CODE REVIEW

**Comment by kohenkatz** [comment]
The one open question I have is whether it is better to "opt-out" for Windows (as I have done) or to "opt-in" for known-supported Linux and macOS.

On one hand, I would assume that this works on the *BSD platforms, so "opt-in" would be a breaking change for them. On the other hand, it doesn't seem like anyone is complaining about any other platform (though the only Go platform as yet unmentioned that I know of is Solaris/illumos).

**Comment by sagikazarmark** [comment]
From an application point of view I would expect that this library will just transparently _not work_ on windows, meaning the application should not have to handle windows specific branches (is it what `ErrNotSupported` is for?).

Not sure what the current implementation does, my 2 cents only.

**Comment by kohenkatz** [comment]
@sagikazarmark no, it is not handled transparently. The point of returning `ErrNotSupported` is to inform the application that there is no way it can work. It then becomes the responsibility of the application developer to handle that error. One of the possible ways that the developer can do that is by using the stub implementation provided here. However, the point is to allow the developer to choose what happens in the event of running on an unsupported system.

File: env_windows.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package tableflip
+
+import (
+	"os"
+)
+
+// replace Unix-specific syscall with a no-op so it will build
+// without errors.
+
+var stdEnv = &env{
```
Comment (lmb) [inline env_windows.go:10]: This could be just `nil`?

File: errors.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package tableflip
+
+type ErrNotSupported struct{}
```
Comment (lmb) [inline errors.go:3]: I think it's enough to make this
```
var ErrNotSupported = errors.New("tableflip: platform does not support graceful restart")
```

File: testing/upgrader.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+// However, the function still needs to be callable without errors
+// in order to be useful.
+func (u *Upgrader) Ready() error {
+	return nil
+}
+
+// Exit returns a channel which is closed when the process should
+// exit.
+func (u *Upgrader) Exit() <-chan struct{} {
+	return u.exitC
```
Comment (lmb) [inline testing/upgrader.go:47]: You can return nil here, reading from a nil channel always blocks.

File: testing/http_example_test.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	})
+	if errors.Is(err, tableflip.ErrNotSupported{}) {
+		upgradeSupported = false
+		upg, _ = testing.New()
+	} else if err != nil {
+		panic(err)
+	}
+	defer upg.Stop()
+
+	if upgradeSupported {
```
Comment (lmb) [inline testing/http_example_test.go:53]: Would the code fail if you skipped this check? It'd be nice if the only modification required would be to do `upg = testing.New()`.

File: upgrader.go [hunk 5]
```diff
@@ -303,3 +312,10 @@ func writePIDFile(path string) error {
 
 	return os.Rename(fh.Name(), path)
 }
+
+// Check if this is a supported OS.
+// That is currently all Unix-like OS's.
+// At the moment, we assume that is everything except Windows.
+func isSupportedOS() bool {
+	return runtime.GOOS != "windows"
```
Comment (lmb) [inline upgrader.go:320]: This is probably fine.

File: upgrader.go [hunk 6]
```diff
@@ -170,6 +175,10 @@ func (u *Upgrader) HasParent() bool {
 
 // Upgrade triggers an upgrade.
 func (u *Upgrader) Upgrade() error {
+	if !isSupportedOS() {
```
Comment (lmb) [inline upgrader.go:178]: Is this required if `isSupportedOS()` is checked in New?

File: testing/upgrader.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	return nil
+}
+
+// HasParent is always false, since the stub implementation can never
+// have a parent
+func (u *Upgrader) HasParent() bool {
+	return false
+}
+
+func (u *Upgrader) Upgrade() error {
```
Comment (lmb) [inline testing/upgrader.go:67]: Document that this always returns an error?

Comment (lmb) [reply testing/upgrader.go:67]: Maybe even return ErrNotSupported?

File: testing/upgrader.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+type Upgrader struct {
+	*Fds
+
+	exitC chan struct{}
+}
+
+// New creates a new stub Upgrader.
+//
+// Unlike the real version, this can be called many times.
+func New() (*Upgrader, error) {
```
Comment (lmb) [inline testing/upgrader.go:23]: How do you feel about just using `Upgrader{}` in calling code? Without the channel you don't really need New.

File: testing/fds.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
+package testing
+
+import (
+	"net"
+	"os"
+)
+
+type Fds struct{}
```
Comment (lmb) [inline testing/fds.go:8]: I think it's OK to merge testing.Fds and testing.Upgrader into a single struct. User code will always want to use them as a single interface, no?

File: errors.go (hunk 2 above)
Comment (sagikazarmark) [reply errors.go:3]: Global variables unfortunately can easily be changed.

I would suggest something like this: https://dave.cheney.net/2016/04/07/constant-errors

**Comment by lmb** [comment]
> From an application point of view I would expect that this library will just transparently not work on windows, meaning the application should not have to handle windows specific branches (is it what ErrNotSupported is for?).

I don't really want to get into the business of defining what "transparently not work" means TBH. tableflip isn't cross platform, and your application will have to deal with that in some way. With @kohenkatz's changes, code completion and compiling will work out of the box. If you want to be able to also run your binary we've got a ready made stub for you. If you later decide that you want a Windows specific implementation of this, you can easily incorporate it by checking for `ErrNotSupported`. You can also come up with your own stub, that better suits your needs. Finally, we don't have to litter all public functions of Upgrader with calls to `isSupportedOS`.

File: errors.go (hunk 2 above)
Comment (lmb) [reply errors.go:3]: That's true, however I've never found this to be a problem really. My main intention is that the error works with errors.Is like in https://blog.golang.org/go1.13-errors. Up to @kohenkatz.

**Comment by sagikazarmark** [comment]
Okay, let me rephrase it: how would an application look like that needs to run on Windows (for development purposes), but is primarily targeted at unix platforms?

Do I have to use a different instance on windows?

> Finally, we don't have to litter all public functions of Upgrader with calls to isSupportedOS

That's kinda what platform build tags are for IMHO.

I agree that the *behavior* is not cross platform, and I think it's also fine if the final upgrade call exits with an error (the application is shutting down anyway), but listeners and everything else could just work the same way.

There is a chance I don't see the whole picture, but a not supported error and a separate stub all suggest that I will have to use a different instance based on whether I'm on windows or not which could already be decided build time by the library. I think it would even be okay to have a constructor function with different implementations on different platforms.

File: errors.go (hunk 2 above)
Comment (kohenkatz) [reply errors.go:3]: The benefit of doing it the way @sagikazarmark describes is that you can add [Dave Cheney's other error handling ideas](https://dave.cheney.net/2014/12/24/inspecting-errors), but this project isn't using those for its existing errors, so I don't think it really matters either way.

Since I'm making changes anyway, I'm going to change it as originally requested just to be consistent with errors already in the project.

File: testing/http_example_test.go (hunk 4 above)
Comment (kohenkatz) [reply testing/http_example_test.go:53]: Technically it isn't required, since Windows cannot generate `SIGHUP`.
The reason I added it is because there's no reason to be running the useless goroutine.
So, no, it won't fail if you skip this check.
If you'd like I can remove the `if` and just put in a comment describing it for those who want to put it back.

File: upgrader.go (hunk 6 above)
Comment (kohenkatz) [reply upgrader.go:178]: Good point. Removing.

File: testing/upgrader.go (hunk 8 above)
Comment (kohenkatz) [reply testing/upgrader.go:23]: The only reason I'd leave it like this is because it looks the same as the actual `Upgrader`. I think the visual symmetry between the methods is better. That way we don't need to confuse people with two different ways of doing it.

Compare:

```
upg, err := tableflip.New(tableflip.Options{
	PIDFile: *pidFile,
})
if errors.Is(err, tableflip.ErrNotSupported) {
	upg, _ = testing.New()
} else if err != nil {
	panic(err)
}
```

vs

```
upg, err := tableflip.New(tableflip.Options{
	PIDFile: *pidFile,
})
if errors.Is(err, tableflip.ErrNotSupported) {
	upg = &testing.Upgrader{}
} else if err != nil {
	panic(err)
}
```

File: testing/fds.go (hunk 9 above)
Comment (kohenkatz) [reply testing/fds.go:8]: There are some people who [call `Upgrader.Fds.Listen()`](https://github.com/sagikazarmark/modern-go-application/blob/64673035cc3ceead255e1cd675f46a8febaeb53a/cmd/modern-go-application/main.go#L198), and this way will allow them to keep doing that if they want. (It means the interface they have to write will be more complicated, but ensures backwards compatibility.)

File: env_windows.go (hunk 1 above)
Comment (kohenkatz) [reply env_windows.go:10]: It won't compile if it is `nil`

Comment (kohenkatz) [reply env_windows.go:10]: Yes, it can

File: testing/upgrader.go (hunk 8 above)
Comment (lmb) [reply testing/upgrader.go:22]: Well, you can't give testing.New() any options, and it also never returns an error? I'm not too fussed about this though.

**Comment by kohenkatz** [comment]
> how would an application look like that needs to run on Windows (for development purposes), but is primarily targeted at unix platforms?

The minimum required change is defining an interface and adding two lines of code:

**Before**:

```
var upg tableflip.Upgrader
upg, err := tableflip.New(tableflip.Options{
	PIDFile: *pidFile,
})
if err != nil {
	panic(err)
}
```

**After**:

```
type upgrader interface {
	Listen(network, addr string) (net.Listener, error)
	Stop()
	Upgrade() error
	Ready() error
	Exit() <-chan struct{}
}
```

and

```
var upg upgrader
upg, err := tableflip.New(tableflip.Options{
	PIDFile: *pidFile,
})
if errors.Is(err, tableflip.ErrNotSupported) {
	upg, _ = testing.New()
} else if err != nil {
	panic(err)
}
```

**Comment by lmb** [comment]
> There is a chance I don't see the whole picture, but a not supported error and a separate stub all suggest that I will have to use a different instance based on whether I'm on windows or not which could already be decided build time by the library.

Yes, you'll have to define an interface and then use testing.Upgrader when encountering ErrNotSupported. Based on the example in this PR it's about 10 lines of code. I don't think that is too onerous. Do you?

> I think it would even be okay to have a constructor function with different implementations on different platforms.

Then you as a user can't discern whether you're on a supported platform without trying to do an upgrade. Not great.

I see these benefits with the current approach:
1. Users that don't use or develop on Windows don't have to deal with Windows support
2. Users can discern whether they are on a supported platform
3. Based on this they can either stub out the library
4. Or use a different library that works on windows

Using build tags means 3 and 4 are not straightforward. As a bonus, the current approach is useful if you want to unit test your main function: in tests you can pass in testing.Upgrader or your own implementation to make sure your code behaves as expected.

File: testing/upgrader.go (hunk 8 above)
Comment (kohenkatz) [reply testing/upgrader.go:22]: I thought of allowing it to take the same `tableflip.Options` argument and throwing an error if it is actually passed in (since the options aren't supported), but I decided not to bother.

**Comment by lmb** [comment]
Thanks!
